lib/chef/provider/package.rb

```diff
@@ -414,19 +433,19 @@ def target_version_array
             each_package do |package_name, new_version, current_version, candidate_version|
               case action
               when :upgrade
-                if version_equals?(current_version, new_version)
-                  # this is an odd use case
-                  logger.trace("#{new_resource} #{package_name} #{new_version} is already installed -- you are equality pinning with an :upgrade action, this may be deprecated in the future")
-                  target_version_array.push(nil)
-                elsif version_equals?(current_version, candidate_version)
-                  logger.trace("#{new_resource} #{package_name} #{candidate_version} is already installed")
+                if current_version.nil?
```

Comment:
it's entirely unclear here how it's possible for current_version != new_version but current_version != candidate_version. I mean *I* get it because I've been talking to you about those stuff for weeks, but I would add a TON of comments here for future-Lamont and future-Phil.

Reply:
candidate_version and current_version are always fully materialized EVRs (in the yum/dnf subclasses) because they come back from queries.

the new_version comes from the user and can be "1.*" or whatever.  and "1.2-1" != "1.*".  i was going to go down the road of having a `installed_version.satisfied_by?(new_version)` API that could be called in the subclass, but that looked like it was going to require a lot of python code that would duplicate internal code in the dnf library.  making it so that the installed_version was just automatically satisfied_by the new_version (or else it being nil) was vastly simpler and will be more robust and gets what we actually need.  that way if either candidate_version or installed_version is nil it either doesn't exist, or doesn't satisfy the desired constraint and either way we have to do work to fix it.

Reply:
Oh I understand it. I'm just saying .... comment your damn code. :)